File: 01-arrays_&_hashing/00347-Top_K_Frequent_Elements/00347-top_k_frequent_elements.cpp
// 347: Top K Frequent Elements
// https://leetcode.com/problems/top-k-frequent-elements/


#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    // SOLUTION
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> tmap;
        for (int num : nums) ++tmap[num];
        
        vector<vector<int>> buckets(nums.size() + 1); 
        for (auto p : tmap) buckets[p.second].push_back(p.first);
        
        vector<int> result;
        for (int i=buckets.size()-1; i>=0 && result.size()<k; --i) {
            for (int num : buckets[i]) {
                result.push_back(num);
                if (result.size() == k) break;
            }
        }

        return result;
    }
};


int main() {
    Solution o;

    // INPUT
    vector<int> nums = {1,1,1,2,2,3};
    int k = 2;

    // OUTPUT
    auto result = o.topKFrequent(nums, k);
    cout << "["; for (auto v : result) cout << v << ","; cout << "\b]" << endl;
    
    return 0;
}

File: 01-arrays_&_hashing/00347-Top_K_Frequent_Elements/00347-top_k_frequent_elements.java
// 347: Top K Frequent Elements
// https://leetcode.com/problems/top-k-frequent-elements/


import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;


class Solution {
    // SOLUTION
    public int[] topKFrequent(int[] nums, int k) {
        List<Integer>[] buckets = new List[nums.length + 1];
        Map<Integer, Integer> tmap = new HashMap<Integer, Integer>();
    
        for (int n : nums) {
            tmap.put(n, tmap.getOrDefault(n, 0) + 1);
        }
    
        for (int key : tmap.keySet()) {
            int frequency = tmap.get(key);
            if (buckets[frequency] == null) {
                buckets[frequency] = new ArrayList<>();
            }
            buckets[frequency].add(key);
        }
    
        List<Integer> result = new ArrayList<>();
        for (int pos = buckets.length - 1; pos >= 0 && result.size() < k; pos--) {
            if (buckets[pos] != null) {
                result.addAll(buckets[pos]);
            }
        }

        return result.stream().mapToInt(i -> i).toArray();
    }

    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        int nums[] = {1,1,1,2,2,3};
        int k = 2;

        // OUTPUT
        var result = o.topKFrequent(nums, k);
        System.out.print("["); for (var v : result) System.out.print(v+","); System.out.println("\b]");
    }
}
